### PR TITLE
Implemented basic version of AOTDispatcher that only chooses between autograd or no autograd

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83367
* #83361
* __->__ #83248

### CODE REVIEW

File: functorch/functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
     """
-    Traces the forward and backward graphs of the attr:`flat_fn` to generate a
-    joint graph. The joint graph is an Fx graph with Aten ops. Please refer to
-    the tracing mechanism to understand the graph capturing details.
 
-    The joint graph is then passed through attr:`partition_fn` to isolate the
-    forward and backward portions, which are then respectively compiled via the
-    provided attr:`fw_compiler` and attr:`bw_compiler`.
+    fw_compiler: Callable
+    bw_compiler: Callable
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:204]: Doesn't need to happen in this PR, but as discussed offline it is a bit strange to have both "fw_compiler" and "bw_compiler" if we are in a situation with higher order gradients

File: functorch/functorch/_src/compilers.py [hunk 2]
```diff
@@ -142,6 +142,7 @@ def nnc_jit(f, static_argnums=None):
 default_decompositions = get_decompositions(default_decompositions)
 
 
+@make_boxed_compiler
```
Comment (zou3519) [inline functorch/functorch/_src/compilers.py:145]: Just checking -- this changes the semantics of print_compile, right?

File: functorch/functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
+    Traces the forward and backward graphs of the attr:`flat_fn` to generate a
+    joint graph. The joint graph is an Fx graph with Aten ops. Please refer to
+    the tracing mechanism to understand the graph capturing details.
+
+    The joint graph is then passed through attr:`partition_fn` to isolate the
+    forward and backward portions, which are then respectively compiled via the
+    provided attr:`fw_compiler` and attr:`bw_compiler`.
+
+    The resulting compiled forward and backward graphs are then wrapped up in a
+    ``torch.autograd.Function`` object.
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:321]: Docstring no longer correct if tensors don't require true or grad is not enabled, right?

File: functorch/functorch/_src/aot_autograd.py [hunk 4]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
 
-            torch._C._jit_set_autocast_mode(old_jit_autocast_flag)
             return tuple(out)
 
-    return CompiledFunction
+    return CompiledFunction.apply
+
+
+def create_aot_dispatcher_function(
+    flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, grad_enabled: bool
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:309]: Do we need to pass grad_enabled? Can we just query it inside the function?

File: functorch/functorch/_src/aot_autograd.py [hunk 5]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
+                else x,
+                flat_args,
+            )
+            fake_flat_tensor_args = pytree.tree_map(
+                lambda x: mode.from_tensor(x)
+                if mode
+                else x
+                if isinstance(x, Tensor)
+                else x,
+                flat_args,
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:347]: Lol I really can't parse this code

File: functorch/functorch/_src/aot_autograd.py [hunk 6]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
-    ``torch.autograd.Function`` object.
-    """
-    if decompositions is None:
-        decompositions = {}
+
+def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
+    def fake_signature(fn, nargs):
+        """FX gets confused by varargs, de-confuse it"""
+        argnames = ",".join(f"arg{i}" for i in range(nargs))
+        return eval(f"lambda {argnames}: fn({argnames})", {"fn": fn})
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:213]: fake_signature is already defined above

Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:213]: for my understanding: the way this is de-confusing it is... passing in the args one at a time? The eval feels hacky, how difficult would it be to teach FX about varargs?

File: functorch/functorch/_src/aot_autograd.py [hunk 7]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
     if not steal_args:
         args = list(args)
     assert isinstance(args, list)
-    # TODO: Please remove soon
-    # https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670
+
+    # Disable the JIT Autocast flag to prevent re-autocasting of jitted graph.
+    # TODO - Remove when https://github.com/pytorch/functorch/pull/794 is fixed.
+    old_jit_autocast_flag = torch._C._jit_set_autocast_mode(False)
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:181]: this should be a try/catch or a context manager, if something fails before restoring the autocast mode state then the function is left in a weird state

File: functorch/functorch/_src/aot_autograd.py [hunk 8]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
-                    compiled_bw = bw_compiler(bw_module, bw_args)
-            else:
-                fw_outs = call_func_with_args(compiled_fw, flat_tensor_args)
-            torch._C._jit_set_autocast_mode(old_jit_autocast_flag)
+    class CompiledFunction(torch.autograd.Function):
+        @staticmethod
+        @disable_torchdynamo
+        def forward(ctx, *flat_tensor_args):
+            # Disable the JIT Autocast flag to prevent re-autocasting of jitted graph.
+            # TODO - Remove when https://github.com/pytorch/functorch/pull/794 is fixed.
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:290]: First line of comment doesn't seem relevant anymore? autocast flag is being disabled inside the function

**Review by zou3519 — APPROVED** [review]
Seems reasonable. I assume most of this is a refactor (so a lot of the minor comments I had apply to the original code), and the new part of it looks like the aot_dispatcher_base path where autograd isn't involved.
- Main point from me is a question about if the aot_dispatcher API should just take in a single `compiler` arg instead of fw_compiler / bw_compiler
- Would this benefit from some explicit tests for case where we don't construct autograd.Function?
- Some minor comments around context managers that save and restore state
- Some of the docstrings are now out of date

File: functorch/functorch/_src/aot_autograd.py (hunk 5 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:347]: lol blame black :P

File: functorch/functorch/_src/aot_autograd.py [hunk 9]
```diff
@@ -10,7 +11,7 @@
 from torch import Tensor
 from torch._subclasses import FakeTensorMode
 from torch.fx import immutable_collections
-from torch.nn.utils import _stateless
+from torch.nn.utils import _stateless, stateless
```
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:14]: fix lint plz

File: functorch/functorch/_src/aot_autograd.py [hunk 10]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
-    joint graph. The joint graph is an Fx graph with Aten ops. Please refer to
-    the tracing mechanism to understand the graph capturing details.
 
-    The joint graph is then passed through attr:`partition_fn` to isolate the
-    forward and backward portions, which are then respectively compiled via the
-    provided attr:`fw_compiler` and attr:`bw_compiler`.
+    fw_compiler: Callable
+    bw_compiler: Callable
+    partition_fn: Callable
+    decompositions: Dict[Callable, Callable]
```
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:206]: It would be nice to have more detailed annotations for these functions, it would not be difficult to do

File: functorch/functorch/_src/aot_autograd.py (hunk 6 above)
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:213]: not necessary on master, this is built into `make_fx` now

File: functorch/functorch/_src/aot_autograd.py [hunk 11]
```diff
@@ -174,146 +175,192 @@ def call_func_with_args(f, args, steal_args=False):
...
+    )(*flat_args)
+    with track_graph_compiling("forward"):
+        compiled_fw = aot_config.fw_compiler(fw_module, flat_args)
+
+    @wraps(compiled_fw)
+    def new_fn(args):
+        fw_outs = call_func_with_args(compiled_fw, args)
+        return fw_outs
+
+    return new_fn
```
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:226]: `return partial(call_func_with_args, compiled_fw)`

File: functorch/functorch/_src/aot_autograd.py [hunk 12]
```diff
@@ -351,6 +398,12 @@ def unflatten(self, x):
         return pytree.tree_unflatten(x, self.spec)
 
 
+def fake_signature(fn, nargs):
+    """FX gets confused by varargs, de-confuse it"""
+    argnames = ",".join(f"arg{i}" for i in range(nargs))
+    return eval(f"lambda {argnames}: fn({argnames})", {"fn": fn})
```
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:404]: second copy lol

**Review by ezyang — APPROVED** [review]

File: functorch/test/test_pythonkey.py [hunk 13]
```diff
@@ -266,7 +266,7 @@ def assert_graph_empty(fx_g, _):
         f = aot_function(foo, nop, assert_graph_empty)
         with torch.set_grad_enabled(False):
             f(*inps)
-        self.assertEqual(graph_size, 2)
+        self.assertTrue(graph_size is None)
```
Comment (ezyang) [inline functorch/test/test_pythonkey.py:269]: assertIsNone

File: functorch/functorch/_src/aot_autograd.py (hunk 9 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:15]: it's the damn "format everything with black commit lower down in the stack" >:( It makes rebasing past that so annoying every time.

I've learned my lesson about putting autoformat commits in the middle of stacks lol

File: functorch/functorch/_src/compilers.py (hunk 2 above)
Comment (Chillee) [reply functorch/functorch/_src/compilers.py:145]: I don't think so. What semantics do you think it's changing?

File: functorch/functorch/_src/aot_autograd.py (hunk 4 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:310]: hmm... I think you're right. We originally needed this since we were compiling inside an autograd.Function for no reason and the autograd state is always disabled in there.

File: functorch/functorch/_src/aot_autograd.py (hunk 1 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:205]: This is true in principle, but there's a decent amount of folks using `fw_compiler` and `bw_compiler` separately right now - for example, cudagraphs integrations benefit from this.

File: functorch/functorch/_src/aot_autograd.py (hunk 3 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:322]: Technically true, but I don't think this decision is really that user facing.

File: functorch/functorch/_src/aot_autograd.py (hunk 7 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:182]: We're worried about overhead of a context manager. Also, I think the case we're catching is pretty unusual, and would require a pretty unusual confluence of situations to matter (torchscript + aotautograd in same program, amp, aotautograd errors inside call but we want to recover, etc.)

Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:182]: Actually, seems like you only need this during compilation.
